core/staking_verifier.go

```diff
@@ -0,0 +1,279 @@
+}
+
+// VerifyAndUndelegateFromMsg verifies the undelegate validator message
+// using the stateDB & chainContext and returns the edited validatorWrapper
+// with the undelegation applied to it.
+//
+// Note that this function never updates the stateDB, it only reads from stateDB.
+func VerifyAndUndelegateFromMsg(
+	stateDB vm.StateDB, epoch *big.Int, msg *staking.Undelegate,
+) (*staking.ValidatorWrapper, error) {
+	if stateDB == nil {
+		return nil, errStateDBIsMissing
+	}
+	if epoch == nil {
+		return nil, errEpochMissing
+	}
+	if msg.Amount.Sign() == -1 {
+		return nil, errNegativeAmount
+	}
+	if !stateDB.IsValidator(msg.ValidatorAddress) {
+		return nil, errValidatorNotExist
+	}
+	wrapper := stateDB.GetStakingInfo(msg.ValidatorAddress)
+	if wrapper == nil {
+		return nil, errValidatorNotExist
+	}
+	for i := range wrapper.Delegations {
+		delegation := &wrapper.Delegations[i]
+		if bytes.Equal(delegation.DelegatorAddress.Bytes(), msg.DelegatorAddress.Bytes()) {
+			if err := delegation.Undelegate(epoch, msg.Amount); err != nil {
+				return nil, err
+			}
+			if err := wrapper.SanityCheck(); err != nil {
+				return nil, err
+			}
+			return wrapper, nil
+		}
+	}
+	return nil, errNoDelegationToUndelegate
+}
+
+// VerifyAndCollectRewardsFromDelegation verifies and collects rewards
+// from the given delegation slice using the stateDB. It returns all of the
+// edited validatorWrappers and the sum total of the rewards.
+//
+// Note that this function never updates the stateDB, it only reads from stateDB.
+func VerifyAndCollectRewardsFromDelegation(
+	stateDB vm.StateDB, delegations []staking.DelegationIndex,
+) ([]*staking.ValidatorWrapper, *big.Int, error) {
+	if stateDB == nil {
+		return nil, nil, errStateDBIsMissing
+	}
+	updatedValidatorWrappers := []*staking.ValidatorWrapper{}
+	totalRewards := big.NewInt(0)
+	for i := range delegations {
+		delegation := &delegations[i]
+		wrapper := stateDB.GetStakingInfo(delegation.ValidatorAddress)
+		if wrapper == nil {
+			return nil, nil, errValidatorNotExist
+		}
+		if uint64(len(wrapper.Delegations)) > delegation.Index {
+			delegation := &wrapper.Delegations[delegation.Index]
+			if delegation.Reward.Cmp(big.NewInt(0)) > 0 {
+				totalRewards.Add(totalRewards, delegation.Reward)
+			}
+			delegation.Reward.SetUint64(0)
+		}
+		if err := wrapper.SanityCheck(); err != nil {
+			return nil, nil, err
+		}
+		updatedValidatorWrappers = append(updatedValidatorWrappers, wrapper)
+	}
+	if totalRewards.Int64() == 0 {
+		return nil, nil, errNoRewardsToCollect
+	}
+	return updatedValidatorWrappers, totalRewards, nil
+}
+
+// getPendingEpoch returns the epoch of the next block given a chain
+func getPendingEpoch(chain interface{ CurrentBlock() *types.Block }) *big.Int {
+	currentBlock := chain.CurrentBlock()
+	pendingBlockNumber := new(big.Int).Add(currentBlock.Number(), big.NewInt(1))
+	blocksPerEpoch := new(big.Int).SetUint64(
+		shard.Schedule.InstanceForEpoch(currentBlock.Epoch()).BlocksPerEpoch(),
```

Comment:
actually, don't add custom func like this. Just use the existing util "shard.Schedule.IsLastBlock". It will tell whether it's the last block of an epoch. From IsLastBlock, you can see the epoch calculation is not that straightforward due to some legacy nuances.

Reply:
oh alright, corrected for it.